### PR TITLE
Reverse color list - Related to issue #285

### PR DESCRIPTION
**Pull request by martinmanzo**
## Description

There's a new attribute for bar charts that set whether or not the color list should be reversed. By default it is set to true to make the change backward compatible.

## Motivation and Context

https://github.com/eventbrite/britecharts/issues/285

## How Has This Been Tested?

I ran the tests included before and after introducing the new code. Then played with the demos a bit to see if it worked ok.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project. --> I think it does! UPDATED: It does.
- [ ] My change requires a change to the documentation. --> I updated it by running npm run docs.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes. --> I'm not D3 savy enough to know how to test the change. Will do it if necessary. UPDATED: Added one test.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/charts/bar.js [hunk 1]
```diff
@@ -100,42 +102,64 @@ define(function(require) {
...
-
-            _percentageLabelHorizontalX = ({value}) => xScale(value) + percentageLabelMargin,
-            _percentageLabelHorizontalY= ({name}) => yScale(name) + (yScale.bandwidth() / 2) + (percentageLabelSize * (3/8)),
-
-            _percentageLabelVerticalX = ({name}) => xScale(name),
-            _percentageLabelVerticalY = ({value}) => yScale(value) - percentageLabelMargin,
-
-            _percentageLabelHorizontalFormatValue = ({value}) => d3Format.format(horizontalLabelFormat)(value),
-            _percentageLabelVerticalFormatValue = ({value}) => d3Format.format(verticalLabelFormat)(parseFloat(value) * 100);
+            getName = ({
```
Comment (Golodhros) [inline src/charts/bar.js]: I guess you have a linter that is unfolding these automatically? we prefer it as they were. Thanks!

File: src/charts/bar.js [hunk 2]
```diff
@@ -173,9 +197,9 @@ define(function(require) {
          * Also applies the Margin convention
          * @private
          */
-        function buildContainerGroups(){
+        function buildContainerGroups() {
             let container = svg
-              .append('g')
+                .append('g')
```
Comment (Golodhros) [inline src/charts/bar.js]: That 2 spaces indentation has some logic behind. It is a d3 pattern useful to mark when a selection changes.

File: src/charts/bar.js [hunk 3]
```diff
@@ -220,25 +244,39 @@ define(function(require) {
...
          * Builds the SVG element that will contain the chart
          * @param  {HTMLElement} container DOM element that will work as the container of the graph
          * @private
          */
-        function buildSVG(container){
+        function buildSVG(container) {
             if (!svg) {
                 svg = d3Selection.select(container)
-                  .append('svg')
+                    .append('svg')
```
Comment (Golodhros) [inline src/charts/bar.js]: Same here

File: src/charts/bar.js [hunk 4]
```diff
@@ -220,25 +244,39 @@ define(function(require) {
...
+                    .map(({
+                        name
+                    }, i) => ({
+                        name,
+                        color: colorSchema[i % colorSchema.length]
+                    }));
+            } else {
+                colorList = data.map(d => d)
+                    .map(({
+                        name
```
Comment (Golodhros) [inline src/charts/bar.js]: Let's have these in one line

File: src/charts/bar.js [hunk 5]
```diff
@@ -302,29 +343,41 @@ define(function(require) {
         function drawHorizontalBars(bars) {
             // Enter + Update
             bars.enter()
-              .append('rect')
+                .append('rect')
```
Comment (Golodhros) [inline src/charts/bar.js]: same here

File: src/charts/bar.js [hunk 6]
```diff
@@ -302,29 +343,41 @@ define(function(require) {
...
                 .classed('bar', true)
                 .attr('y', chartHeight)
                 .attr('x', 0)
                 .attr('height', yScale.bandwidth())
-                .attr('width', ({value}) => xScale(value))
-                .attr('fill', ({name}) => colorMap(name))
+                .attr('width', ({
```
Comment (Golodhros) [inline src/charts/bar.js]: same

File: src/charts/bar.js [hunk 7]
```diff
@@ -302,29 +343,41 @@ define(function(require) {
...
                 })
                 .on('mouseout', function() {
                     dispatcher.call('customMouseOut', this);
-                    d3Selection.select(this).attr('fill', ({name}) => colorMap(name))
+                    d3Selection.select(this).attr('fill', ({
+                        name
+                    }) => colorMap(name))
                 })
-              .merge(bars)
+                .merge(bars)
```
Comment (Golodhros) [inline src/charts/bar.js]: same here

**Comment by martinmanzo** [comment]
Hi! Yes I'm using a linter, I'll re check the settings and make this changes later today. Thanks

**Comment by martinmanzo** [comment]
Done! Waiting for new feedback. Thanks!

**Comment by rwholey-eb** [comment]
Hey @martinmanzo! Thanks for your pr! I think you make a good point. Eventually we'd like to come to a place where the order you pass the colors is the order that the colors are shown. We will probably make this change sometime in the near-ish future. This will be a breaking change for this fix.

Question for my own curiosity, in your specific use case, why not reverse your color list before giving it to britecharts? 

I'm okay with merging this pr when the conflicting doc files are resolved! Thanks again :)

**Comment by martinmanzo** [comment]
Hi @rwholey-eb! Thanks for your words.

I tried reversing the color list but it still continued to work wrong. I want to be able to add a legend chart to a bar chart. Because the bar chart reverses the data and the legend chart doesn't an inconsistency is generated. I could try to reproduce the problem again and attach some screenshots if you want!

I'm fixing the docs right now!

**Comment by rwholey-eb** [comment]
awesome! thanks! No need for screenshots its all good. Also if those conflicts get a bit weird sometimes i'll rm the whole docs file and run npm run docs to rebuild them. It's also on our roadmap to rework the way our project is built so we dont have all those file changes with each pr! What a great day that will be....

**Comment by martinmanzo** [comment]
Yeah I merged with the last version of Britecharts and rebuilt the docs, they were solved. I decided to write a test. I'll push when ready!

BTW leave the way the project is built as "technical debt" :see_no_evil: :smile:

**Comment by rwholey-eb** [comment]
@Golodhros you still have changes requested, care to take a second look before I merge?

**Comment by Golodhros** [comment]
Good to go! Thanks!

**Pull request merged**